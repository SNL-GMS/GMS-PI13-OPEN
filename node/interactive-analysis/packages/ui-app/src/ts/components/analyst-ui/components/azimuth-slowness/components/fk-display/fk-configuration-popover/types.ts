import { FkTypes } from '@gms/common-graphql';
import { CheckboxSearchListTypes } from '@gms/ui-core-components';
import { FkConfigurationWithUnits, FkUnits } from '~analyst-ui/components/azimuth-slowness/types';

export enum FkConfigurationPopoverPanel {
  DEFAULT,
  ADVANCED
}

export interface FkConfigurationPopoverState {
  openPanel: FkConfigurationPopoverPanel;
  fkUnits: FkUnits;
  normalizeWaveforms: boolean;
  mediumVelocity: number;
  maximumSlowness: number;
  numberOfPoints: number;
  channelCheckboxes: CheckboxSearchListTypes.CheckboxItem[];
  useVerticalChannelOffsets: boolean;
}
export interface FkConfigurationPopoverProps extends FkTypes.FkConfiguration {
  fkUnitDisplayed: FkUnits;
  applyFkConfiguration(configuration: FkConfigurationWithUnits);
  close();
}
