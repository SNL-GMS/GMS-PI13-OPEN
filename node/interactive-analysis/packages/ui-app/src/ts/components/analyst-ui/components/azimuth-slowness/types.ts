import {
  CommonTypes,
  ConfigurationTypes,
  EventTypes,
  FkTypes,
  ProcessingStationTypes,
  SignalDetectionTypes,
  WaveformTypes
} from '@gms/common-graphql';
import GoldenLayout from '@gms/golden-layout';
import { Client } from '@gms/ui-apollo';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import Immutable from 'immutable';
import { ChildProps, MutationFunction } from 'react-apollo';
import { FilterType, FkThumbnailSize } from './components/fk-thumbnail-list/fk-thumbnails-controls';
import { LeadLagPairs } from './constants';

export interface FkParams {
  windowParams: FkTypes.WindowParameters;
  frequencyPair: FkTypes.FrequencyBand;
}

export interface LeadLagPairAndString {
  leadLagPairs: LeadLagPairs;
  windowParams: FkTypes.WindowParameters;
}

export enum FkUnits {
  FSTAT = 'FSTAT',
  POWER = 'POWER'
}

/**
 * Used to return a super set of the fk configuration from the fk config popover
 */
export interface FkConfigurationWithUnits extends FkTypes.FkConfiguration {
  fkUnitToDisplay: FkUnits;
}

/**
 * Azimuth Slowness Redux Props
 */
export interface AzimuthSlownessReduxProps {
  // passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  client: Client;
  currentStageInterval: AnalystWorkspaceTypes.StageInterval;
  currentTimeInterval: CommonTypes.TimeRange;
  selectedSdIds: string[];
  openEventId: string;
  sdIdsToShowFk: string[];
  analystActivity: AnalystWorkspaceTypes.AnalystActivity;
  location: AnalystWorkspaceTypes.LocationSolutionState;
  defaultSignalDetectionPhase?: CommonTypes.PhaseType;
  channelFilters: Immutable.Map<string, WaveformTypes.WaveformFilter>;
  selectedSortType: AnalystWorkspaceTypes.WaveformSortType;
  setSelectedSdIds(ids: string[]): void;
  setChannelFilters(filters: Immutable.Map<string, WaveformTypes.WaveformFilter>): void;
  setSdIdsToShowFk(signalDetectionIds: string[]): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
}

export interface SubscriptionAction {
  (
    list: SignalDetectionTypes.SignalDetection[],
    index: number,
    prev: SignalDetectionTypes.SignalDetection[],
    currentIteree: SignalDetectionTypes.SignalDetection
  ): void;
}

/**
 * Azimuth Slowness State
 */
export interface AzimuthSlownessState {
  fkThumbnailSizePx: FkThumbnailSize;
  fkThumbnailColumnSizePx: number;
  filterType: FilterType;
  userInputFkWindowParameters: FkTypes.WindowParameters;
  userInputFkFrequency: FkTypes.FrequencyBand;
  numberOfOutstandingComputeFkMutations: number;
  fkUnitsForEachSdId: Immutable.Map<string, FkUnits>;
  fkInnerContainerWidthPx: number;
  fkFrequencyThumbnails: Immutable.Map<string, FkTypes.FkFrequencyThumbnail[]>;
}

/**
 * Mutations used by the Az Slow display
 */
export interface AzimuthSlownessMutations {
  computeFks: MutationFunction<{}>;
  computeFkFrequencyThumbnails: MutationFunction<{}>;
  setWindowLead: MutationFunction<{}>;
  markFksReviewed: MutationFunction<{}>;
}

/**
 * Consolidated props for Azimuth Slowness
 */
export type AzimuthSlownessProps = AzimuthSlownessReduxProps &
  ChildProps<AzimuthSlownessMutations> &
  ProcessingStationTypes.DefaultStationsQueryProps &
  ConfigurationTypes.UIConfigurationQueryProps &
  SignalDetectionTypes.SignalDetectionsByStationQueryProps &
  EventTypes.EventsInTimeRangeQueryProps;

/**
 * State of the az slow panel
 */
export interface AzimuthSlownessPanelState {
  currentMovieSpectrumIndex: number;
}

export interface AzimuthSlownessPanelProps {
  // Data
  defaultStations: ProcessingStationTypes.ProcessingStation[];
  eventsInTimeRange: EventTypes.Event[];
  displayedSignalDetection: SignalDetectionTypes.SignalDetection | undefined;
  openEvent: EventTypes.Event | undefined;
  associatedSignalDetections: SignalDetectionTypes.SignalDetection[];
  signalDetectionsToDraw: SignalDetectionTypes.SignalDetection[];
  signalDetectionsIdToFeaturePredictions: Immutable.Map<string, EventTypes.FeaturePrediction[]>;
  signalDetectionsByStation: SignalDetectionTypes.SignalDetection[];
  featurePredictionsForDisplayedSignalDetection: EventTypes.FeaturePrediction[];
  distances: EventTypes.LocationToStationDistance[];
  // Redux state as props
  selectedSdIds: string[];
  defaultWaveformFilters: WaveformTypes.WaveformFilter[];
  sdIdsToShowFk: string[];
  location: AnalystWorkspaceTypes.LocationSolutionState;
  fkFrequencyThumbnails: FkTypes.FkFrequencyThumbnail[];
  // Azimuth display state as props
  fkThumbnailColumnSizePx: number;
  fkDisplayWidthPx: number;
  fkDisplayHeightPx: number;
  filterType: FilterType;
  fkThumbnailSizePx: FkThumbnailSize;
  fkUnitsForEachSdId: Immutable.Map<string, FkUnits>;
  numberOfOutstandingComputeFkMutations: number;
  userInputFkFrequency: FkTypes.FrequencyBand;
  fkUnitForDisplayedSignalDetection: FkUnits;
  userInputFkWindowParameters: FkTypes.WindowParameters;
  fkInnerContainerWidthPx: number;
  channelFilters: Immutable.Map<string, WaveformTypes.WaveformFilter>;
  selectedSortType: AnalystWorkspaceTypes.WaveformSortType;
  defaultSignalDetectionPhase?: CommonTypes.PhaseType;
  // Prop functions
  adjustFkInnerContainerWidth(
    fkThumbnailsContainer: HTMLDivElement,
    fkThumbnailsInnerContainer: HTMLDivElement
  ): void;
  markFksForSdIdsAsReviewed(sdIds: string[]): void;
  updateFkThumbnailSize(size: FkThumbnailSize): void;
  updateFkFilter(filterType: FilterType): void;
  setFkThumbnailColumnSizePx(newSizePx: number): void;
  computeFkAndUpdateState(fkInput: FkTypes.FkInput): void;
  changeUserInputFks(
    windowParams: FkTypes.WindowParameters,
    frequencyBand: FkTypes.FrequencyBand
  ): void;
  setFkUnitForSdId(sdId: string, fkUnit: FkUnits): void;

  // Redux functions
  setSelectedSdIds(sdIds: string[]): void;
  setSdIdsToShowFk(sdIds: string[]): void;
  setChannelFilters(filters: Immutable.Map<string, WaveformTypes.WaveformFilter>): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
}
