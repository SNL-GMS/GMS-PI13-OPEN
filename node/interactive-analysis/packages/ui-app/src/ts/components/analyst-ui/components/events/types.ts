import {
  CommonTypes,
  EventTypes,
  ProcessingStationTypes,
  SignalDetectionTypes
} from '@gms/common-graphql';
import GoldenLayout from '@gms/golden-layout';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import { ChildProps, MutationFunction } from 'react-apollo';
import { gmsColors, semanticColors } from '~scss-config/color-preferences';

/**
 * Types of events which can be show
 */
export enum EventFilters {
  EDGE = 'Edge',
  COMPLETED = 'Completed'
}
export const eventFilterToColorMap: Map<any, string> = new Map();
eventFilterToColorMap.set(EventFilters.EDGE, gmsColors.gmsBackground);
eventFilterToColorMap.set(EventFilters.COMPLETED, semanticColors.analystComplete);

/**
 * Table row object for events
 */
export interface EventsRow {
  id: string;
  eventHypId: string;
  isOpen: boolean;
  stageId: string;
  lat: number;
  lon: number;
  depth: number;
  time: number;
  activeAnalysts: string[];
  numDetections: number;
  status: string;
  edgeEvent: boolean;
  signalDetectionConflicts: SignalDetectionConflict[];
}

export interface SignalDetectionConflict {
  stationName: string;
  id: string;
  arrivalTime: number;
  phase: CommonTypes.PhaseType;
}
/**
 * Event list local state
 */
export interface EventsState {
  currentTimeInterval: CommonTypes.TimeRange;
  suppressScrollOnNewData: boolean;
  showEventOfType: Map<EventFilters, boolean>;
}

/**
 * Mutations used in the event list
 */
export interface EventsMutations {
  updateEvents: MutationFunction<{}>;
  saveEvent: MutationFunction<{}>;
  saveAllModifiedEvents: MutationFunction<{}>;
}

/**
 * Props mapped in from Redux state
 */
export interface EventsReduxProps {
  // Passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  currentTimeInterval: CommonTypes.TimeRange;
  analystActivity: AnalystWorkspaceTypes.AnalystActivity;
  openEventId: string;
  selectedEventIds: string[];

  // callbacks
  setOpenEventId(
    event: EventTypes.Event | undefined,
    latestLocationSolutionSet: EventTypes.LocationSolutionSet | undefined,
    preferredLocationSolutionId: string | undefined
  ): void;
  setSelectedEventIds(ids: string[]): void;
}
export interface SignalDetectionHypothesisWithStation
  extends SignalDetectionTypes.SignalDetectionHypothesis {
  stationName?: string;
}

/**
 * Consolidated props type for event list
 */
export type EventsProps = EventsReduxProps &
  ChildProps<EventsMutations> &
  EventTypes.EventsInTimeRangeQueryProps &
  ProcessingStationTypes.DefaultStationsQueryProps &
  SignalDetectionTypes.SignalDetectionsByStationQueryProps &
  CommonTypes.WorkspaceStateProps;
