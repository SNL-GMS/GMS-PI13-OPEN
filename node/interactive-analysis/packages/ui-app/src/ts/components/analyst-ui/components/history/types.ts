import { CacheTypes, CommonTypes, EventTypes } from '@gms/common-graphql';
import GoldenLayout from '@gms/golden-layout';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import React from 'react';
import { ChildProps, MutationFunction } from 'react-apollo';
import { HistoryEntryAction } from './history-stack/types';

/**
 * Mutations used by the History component
 */
export interface HistoryMutations {
  undoHistory: MutationFunction<CacheTypes.UndoHistoryMutationResult>;
  redoHistory: MutationFunction<CacheTypes.RedoHistoryMutationResult>;
  undoHistoryById: MutationFunction<CacheTypes.UndoHistoryByIdMutationResult>;
  redoHistoryById: MutationFunction<CacheTypes.RedoHistoryByIdMutationResult>;
  undoEventHistory: MutationFunction<CacheTypes.UndoEventHistoryMutationResult>;
  redoEventHistory: MutationFunction<CacheTypes.RedoEventHistoryMutationResult>;
  undoEventHistoryById: MutationFunction<CacheTypes.UndoEventHistoryByIdMutationResult>;
  redoEventHistoryById: MutationFunction<CacheTypes.RedoEventHistoryByIdMutationResult>;
}

/**
 * History Redux Props
 */
export interface HistoryReduxProps {
  // Passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  historyActionInProgress: number;
  openEventId: string;
  currentTimeInterval: CommonTypes.TimeRange;
  analystActivity: AnalystWorkspaceTypes.AnalystActivity;
  keyPressActionQueue: AnalystWorkspaceTypes.KeyAction[];
  setKeyPressActionQueue(actions: AnalystWorkspaceTypes.KeyAction[]): void;
  incrementHistoryActionInProgress();
  decrementHistoryActionInProgress();
}

/**
 * History Props
 */
export type HistoryProps = HistoryReduxProps &
  ChildProps<HistoryMutations> &
  CacheTypes.HistoryQueryProps &
  EventTypes.EventsInTimeRangeQueryProps;

/**
 * State for the HistoryComponent
 */
export interface HistoryState {
  actionIntentPointer: HistoryEntryPointer;
}
/**
 * History Panel Props
 * optional non-ideal-state element to render in the children history-stacks
 */
export interface HistoryPanelProps {
  nonIdealState?: NonIdealHistoryState;
}

/**
 * Keeps track of which HistoryEntryAction is being target for an undo/redo action
 * and provides a check function isIncluded for checking if other actions would be
 * included based on whether the target is an event or global undo/redo.
 */
export interface HistoryEntryPointer {
  entryId: string;
  entryType: HistoryEntryAction;
  isEventMode: boolean;
  isIncluded?(entry: CacheTypes.History): boolean;
  isChangeIncluded(change: CacheTypes.HistoryChange): boolean;
}

/**
 * The type for the data used by history context
 */
export interface HistoryContextData {
  glContainer: GoldenLayout.Container;
  historyList: CacheTypes.History[];
  historyActionIntent: HistoryEntryPointer;
  openEventId: string;
  eventsInTimeRange: EventTypes.Event[];
  historyActionInProgress: number;
  redoEventHistory(quantity?: number): void;
  redoEventHistoryById(id: string): void;
  redoHistory(quantity?: number): void;
  redoHistoryById(id: string): void;
  setHistoryActionIntent(pointer: HistoryEntryPointer): void;
  undoEventHistory(quantity?: number): void;
  undoEventHistoryById(id: string): void;
  undoHistory(quantity?: number): void;
  undoHistoryById(id: string): void;
}

/**
 * History Context to provide access to these high-level props to the History-Component tree.
 * Here, we instantiate the Context variable and set up the defaults for the history context
 */
export const HistoryContext: React.Context<HistoryContextData> = React.createContext<
  HistoryContextData
>({
  glContainer: undefined,
  openEventId: undefined,
  eventsInTimeRange: [],
  undoEventHistory: undefined,
  undoEventHistoryById: undefined,
  redoEventHistory: undefined,
  redoEventHistoryById: undefined,
  undoHistory: undefined,
  undoHistoryById: undefined,
  redoHistory: undefined,
  redoHistoryById: undefined,
  historyActionIntent: undefined,
  setHistoryActionIntent: undefined,
  historyActionInProgress: 0,
  historyList: []
});

export interface NonIdealHistoryState {
  loadingEvent: JSX.Element;
  loadingHistory: JSX.Element;
}
