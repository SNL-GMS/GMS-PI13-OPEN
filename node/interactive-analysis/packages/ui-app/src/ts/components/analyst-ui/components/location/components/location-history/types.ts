import { EventTypes } from '@gms/common-graphql';
import { Row } from '@gms/ui-core-components';
import { AnalystWorkspaceTypes } from '@gms/ui-state';

/**
 * LocationHistory Props
 */
export interface LocationHistoryProps {
  // Current open event
  event: EventTypes.Event;
  location: AnalystWorkspaceTypes.LocationSolutionState;
  setSelectedLocationSolution(locationSolutionSetId: string, locationSolutionId: string): void;
  setSelectedPreferredLocationSolution(
    preferredLocationSolutionSetId: string,
    preferredLocationSolutionId: string
  ): void;
}
/**
 * Location History State
 */
// tslint:disable-next-line: no-empty-interface
export interface LocationHistoryState {}

export interface LocationHistoryRow extends Row {
  locationSolutionId: string;
  locationSetId: string;
  locType: string;
  latestLSSId: string;
  lat: number;
  lon: number;
  depth: number;
  time: string;
  restraint: string;
  smajax: number;
  sminax: number;
  strike: number;
  stdev: number;
  depthRestraintType: EventTypes.DepthRestraintType;
  count: number;
  selectedLocationSolutionSetId: string;
  isLocationSolutionSetPreferred?: boolean;
  isLastInLSSet?: boolean;
  isFirstInLSSet?: boolean;
  preferred?: boolean;
  locationGroup?: LocationHistoryRow[];
  setSelectedPreferredLocationSolution(
    preferredLocationSolutionSetId: string,
    preferredLocationSolutionId: string
  ): void;
  setToSave(locationSolutionSetId: string, locationSolutionId: string): void;
}
