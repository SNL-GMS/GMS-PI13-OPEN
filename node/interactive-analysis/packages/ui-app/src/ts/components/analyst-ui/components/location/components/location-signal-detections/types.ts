import { CommonTypes, EventTypes } from '@gms/common-graphql';
import { Row } from '@gms/ui-core-components';
import Immutable from 'immutable';
import { MutationFunction } from 'react-apollo';
import { SignalDetectionSnapshotWithDiffs } from '../../types';

/**
 * Location Props
 */
export interface LocationSignalDetectionsProps {
  // Current open event
  event: EventTypes.Event;
  // Associated Signal Detections to open event
  signalDetectionDiffSnapshots: SignalDetectionSnapshotWithDiffs[];
  // Distance to the station for selected location
  distances: EventTypes.LocationToStationDistance[];
  // If true, the sd list is showing data that DOES NOT match the rest of the analyst workspace
  historicalMode: boolean;
  // The currently-selected signal detection IDs
  selectedSdIds: string[];

  rejectDetections: MutationFunction<{}>;
  updateDetections: MutationFunction<{}>;
  createEvent: MutationFunction<{}>;
  changeSignalDetectionAssociations: MutationFunction<{}>;

  setSelectedSdIds(ids: string[]): void;
  setSdIdsToShowFk(signalDetectionIds: string[]): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
  showSDContextMenu(selectedSdIds: string[], x: number, y: number);
  showSDDetails(sdId: string, x: number, y: number);
  setDefining(isDefining: boolean, definingType: DefiningTypes): void;
  updateIsDefining(
    definingType: DefiningTypes,
    signalDetectionId: string,
    setDefining: boolean
  ): void;
  toast(message: string): void;
}

export interface LocationSDRow extends Row {
  eventId: string;
  signalDetectionId: string;
  station: string;
  channel: string;
  phase: CommonTypes.PhaseType;
  distance?: number;
  timeObs: string;
  timeRes: number;
  timeCorr: number;
  slownessObs: number;
  slownessRes: number;
  slownessCorr: number;
  azimuthObs: number;
  azimuthRes: number;
  azimuthCorr: number;
  arrivalTimeDefining: boolean;
  slownessDefining: boolean;
  azimuthDefining: boolean;
  // Fields for highlighting if there is a diff between snapshot and state
  isAssociatedDiff: boolean;
  // Removed covers rejected and unassociated
  timeDefiningDiff: boolean;
  azimuthDefiningDiff: boolean;
  slownessDefiningDiff: boolean;
  rejectedOrUnassociated: boolean;
  // Value diffs
  channelNameDiff?: boolean;
  arrivalTimeDiff?: boolean;
  azimuthObsDiff?: boolean;
  slownessObsDiff?: boolean;
  phaseDiff?: boolean;

  historicalMode: boolean;
  updateIsDefining(
    definingType: DefiningTypes,
    signalDetectionHypothesisId: string,
    setDefining: boolean
  );
}

/**
 * DefiningTypes are the FeatureMeasurement Types that populate the Location SD table
 */
export enum DefiningTypes {
  AZIMUTH = 'AZIMUTH',
  SLOWNESS = 'SLOWNESS',
  ARRIVAL_TIME = 'ARRIVAL_TIME'
}
export enum DefiningChange {
  CHANGED_TO_TRUE,
  CHANGED_TO_FALSE,
  CHANGED_TO_OTHER,
  NO_CHANGE
}
/**
 * Describes state of (non)defining sd's
 */
export enum DefiningStates {
  ALL = 'ALL',
  NONE = 'NONE',
  SOME = 'SOME'
}
