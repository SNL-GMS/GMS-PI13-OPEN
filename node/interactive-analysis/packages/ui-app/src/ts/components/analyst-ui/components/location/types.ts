import {
  CommonTypes,
  EventTypes,
  ProcessingStationTypes,
  SignalDetectionTypes
} from '@gms/common-graphql';
import GoldenLayout from '@gms/golden-layout';
import { AnalystWorkspaceTypes } from '@gms/ui-state';
import Immutable from 'immutable';
import { ChildProps, MutationFunction } from 'react-apollo';
import { DefiningChange } from './components/location-signal-detections/types';

export type LocationPanelProps = {
  openEvent: EventTypes.Event;
  signalDetectionsByStation: SignalDetectionTypes.SignalDetection[];
  associatedSignalDetections: SignalDetectionTypes.SignalDetection[];
  distances: EventTypes.LocationToStationDistance[];
  selectedSdIds: string[];
  sdIdsToShowFk: string[];
  measurementMode: AnalystWorkspaceTypes.MeasurementMode;
  widthOfDisplayPx: number;
  location: AnalystWorkspaceTypes.LocationSolutionState;

  setSelectedSdIds(ids: string[]): void;
  setSdIdsToShowFk(signalDetectionIds: string[]): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
  setSelectedLocationSolution(locationSolutionSetId: string, locationSolutionId: string): void;
  setSelectedPreferredLocationSolution(
    preferredLocationSolutionSetId: string,
    preferredLocationSolutionId: string
  ): void;
} & ChildProps<LocationMutations>;

export interface LocationPanelState {
  outstandingLocateCall: boolean;
  sdDefiningChanges: SignalDetectionTableRowChanges[];
}

/**
 * Mutations used by the Location
 */
export interface LocationMutations {
  locateEvent: MutationFunction<{}>;
  createEvent: MutationFunction<{}>;
  updateFeaturePredictions: MutationFunction<{}>;
  updateDetections: MutationFunction<{}>;
  rejectDetections: MutationFunction<{}>;
  changeSignalDetectionAssociations: MutationFunction<{}>;
}

/**
 * Changes to LocationSD table per row
 */
export interface SignalDetectionTableRowChanges {
  signalDetectionId: string;
  // Defining diffs
  arrivalTimeDefining: DefiningChange;
  slownessDefining: DefiningChange;
  azimuthDefining: DefiningChange;
}

/**
 * List of tool tip messages for location button
 */
export enum LocateButtonTooltipMessage {
  NotEnoughDefiningBehaviors = 'minimum defining behaviors must be set',
  Correct = 'Calculates a new event location',
  BadLocationAttributes = 'Location attributes (depth, latitude, or longitude) are not within constraints',
  InvalidData = 'Error in props or query'
}

export interface LocationSDRowDiffs {
  isAssociatedDiff: boolean;
  arrivalTimeDefining: DefiningChange;
  slownessDefining: DefiningChange;
  azimuthDefining: DefiningChange;
  // Value diffs
  channelNameDiff?: boolean;
  arrivalTimeDiff?: boolean;
  azimuthObsDiff?: boolean;
  slownessObsDiff?: boolean;
  phaseDiff?: boolean;
}

export interface DefiningStatus {
  arrivalTimeDefining: DefiningChange;
  slownessDefining: DefiningChange;
  azimuthDefining: DefiningChange;
}

export interface SignalDetectionSnapshotWithDiffs extends EventTypes.SignalDetectionSnapshot {
  diffs: LocationSDRowDiffs;
  rejectedOrUnassociated: boolean;
}

/**
 * Props mapped in from Redux state
 */
export interface LocationReduxProps {
  // Passed in from golden-layout
  glContainer?: GoldenLayout.Container;
  // The currently-open processing interval time range
  currentTimeInterval: CommonTypes.TimeRange;
  // The currently-open event hypothesis IDs
  openEventId: string;
  // The currently-selected signal detection IDs
  selectedSdIds: string[];
  // used for additional time range values
  analystActivity: AnalystWorkspaceTypes.AnalystActivity;
  /** The measurement mode */
  measurementMode: AnalystWorkspaceTypes.MeasurementMode;
  /** The signal detections selected to show fk */
  sdIdsToShowFk: string[];
  /** The selected location state */
  location: AnalystWorkspaceTypes.LocationSolutionState;

  setOpenEventId(
    event: EventTypes.Event | undefined,
    latestLocationSolutionSet: EventTypes.LocationSolutionSet | undefined,
    preferredLocationSolutionId: string | undefined
  ): void;
  setSelectedEventIds(ids: string[]): void;
  setSelectedSdIds(ids: string[]): void;
  setSdIdsToShowFk(signalDetectionIds: string[]): void;
  setMeasurementModeEntries(entries: Immutable.Map<string, boolean>): void;
  setSelectedLocationSolution(locationSolutionSetId: string, locationSolutionId: string): void;
  setSelectedPreferredLocationSolution(
    preferredLocationSolutionSetId: string,
    preferredLocationSolutionId: string
  ): void;
}

/**
 * Location Props
 */
export type LocationProps = LocationReduxProps &
  ChildProps<LocationMutations> &
  ProcessingStationTypes.DefaultStationsQueryProps &
  SignalDetectionTypes.SignalDetectionsByStationQueryProps &
  EventTypes.EventsInTimeRangeQueryProps;
