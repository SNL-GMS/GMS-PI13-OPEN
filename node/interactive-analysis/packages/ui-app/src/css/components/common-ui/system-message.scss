$critical-selector: '[data-severity*=CRITICAL]';
$warning-selector: '[data-severity*=WARNING]';
$info-selector: '[data-severity*=INFO]';

@mixin createColoredCount($selector, $color) {
  #{$selector} &__count {
    color: lighten($color, 10%);
  }
}

.system-message-display {
  @include flexColumn;
  width: 100%;
  height: 100%;
  .toolbar {
    margin-bottom: 0.5em;
  }
}

.system-message-table {
  @include flexColumn;
  width: 100%;
  height: 100%;
  background-color: $gms-prominent-background;
  .system-cell {
    &#{$critical-selector} {
      background-color: $system-message-critical-color;
    }
    &#{$warning-selector} {
      background-color: $system-message-warning-color;
    }
    &#{$info-selector} {
      background-color: $system-message-info-color;
    }
  }
  &__button {
    &--floating {
      position: absolute;
      left: 50%;
      bottom: 3rem;
      transform: translate3D(-50%, -50%, 0);
    }
    &--desc {
      bottom: auto;
      top: 9rem;
    }
  }
}

.system-message-summary {
  @include centerChildren;
  background: $gms-prominent-background;
  height: 100%;
  padding: 0 1em;
  border-radius: $rounded-rectangle-radius;
  font-size: 1rem;

  @include createColoredCount($critical-selector, $gms-strong-warning);
  @include createColoredCount($warning-selector, $gms-warning);
  @include createColoredCount($info-selector, $gms-good-count);

  &__entry {
    @include centerChildren('vertical');
    cursor: pointer;
    & + & {
      margin-left: 1.5em;
    }
    &--disabled .system-message-summary__count {
      color: $gms-soft;
    }
  }

  &__count {
    font-family: $gms-mono;
    font-size: 1.5rem;
    font-weight: bold;
    margin-left: 0.5ch;
    &--prefix {
      opacity: 0.33;
    }
    &--main {
      opacity: 1;
    }
  }
}

/// AG Grid Overrides
.system-message-table {
  .ag-body-viewport,
  .ag-root-wrapper {
    scroll-behavior: smooth;
  }
  .ag-row .ag-cell {
    background-color: lighten($gms-prominent-background, 5%);
    border-left: 1px solid $gms-prominent-background;
    border-right: 1px solid $gms-prominent-background;
  }
  .ag-header-cell {
    padding: 0 10px;
    background-color: $gms-popover-background;
    .ag-header-cell-text {
      color: $gms-main;
    }
  }
  .ag-header-cell-label,
  .ag-header-cell-label {
    padding-left: 7px;
    font-size: 1rem;
    font-weight: 500;
  }
}
