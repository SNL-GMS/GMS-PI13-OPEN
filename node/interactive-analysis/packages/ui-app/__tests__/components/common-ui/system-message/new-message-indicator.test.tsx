import { uuid } from '@gms/common-util';
import * as React from 'react';
import { NewMessageIndicator } from '../../../../src/ts/components/common-ui/components/system-message/new-message-indicator';
import { NewMessageIndicatorProps } from '../../../../src/ts/components/common-ui/components/system-message/types';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

// tslint:disable-next-line: deprecation
const lodash = require.requireActual('lodash');
lodash.uniqueId = () => '1';

let idCount = 0;
uuid.asString = jest.fn().mockImplementation(() => ++idCount);

describe('New Messages Indicator', () => {
  it('is defined', () => {
    expect(NewMessageIndicator).toBeDefined();
  });

  const props: NewMessageIndicatorProps = {
    isVisible: true,
    handleNewMessageIndicatorClick: jest.fn()
  };

  const wrapper = Enzyme.mount(<NewMessageIndicator {...props} />);

  it('matches its snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('scrolls to latest on click', () => {
    wrapper
      .find('[data-cy="new-messages-button"]')
      .first()
      .simulate('click');
    // tslint:disable-next-line: no-unbound-method
    expect(props.handleNewMessageIndicatorClick).toHaveBeenCalledTimes(1);
  });
});
