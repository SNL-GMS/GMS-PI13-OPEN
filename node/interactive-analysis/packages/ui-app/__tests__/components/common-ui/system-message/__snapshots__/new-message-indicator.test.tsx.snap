// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Messages Indicator matches its snapshot 1`] = `
<Component
  handleNewMessageIndicatorClick={[MockFunction]}
  isVisible={true}
>
  <Blueprint3.Button
    className="system-message-table__button system-message-table__button--floating "
    data-cy="new-messages-button"
    icon="arrow-down"
    intent="primary"
    onClick={[Function]}
  >
    <button
      className="bp3-button bp3-intent-primary system-message-table__button system-message-table__button--floating "
      data-cy="new-messages-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      type="button"
    >
      <Blueprint3.Icon
        icon="arrow-down"
        key="leftIcon"
      >
        <span
          className="bp3-icon bp3-icon-arrow-down"
          icon="arrow-down"
        >
          <svg
            data-icon="arrow-down"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              arrow-down
            </desc>
            <path
              d="M13 8c-.3 0-.5.1-.7.3L9 11.6V2c0-.5-.4-1-1-1s-1 .5-1 1v9.6L3.7 8.3C3.5 8.1 3.3 8 3 8c-.5 0-1 .5-1 1 0 .3.1.5.3.7l5 5c.2.2.4.3.7.3s.5-.1.7-.3l5-5c.2-.2.3-.4.3-.7 0-.6-.4-1-1-1z"
              fillRule="evenodd"
              key="0"
            />
          </svg>
        </span>
      </Blueprint3.Icon>
      <span
        className="bp3-button-text"
        key="text"
      >
        New messages
      </span>
      <Blueprint3.Icon
        key="rightIcon"
      />
    </button>
  </Blueprint3.Button>
</Component>
`;
